### PR TITLE
Transfer offset decorations when legalizing laid-out structs

### LINKED ISSUE #5264: WGSL: Nullpointer dereference in WGSLSourceEmitter::emitStructFieldAttributes
Affected tests:

- tests/autodiff/existential-1.slang
- tests/autodiff/existential-2.slang
- tests/autodiff/material/diff-bwd-falcor-material-system.slang
- tests/autodiff/material2/diff-bwd-falcor-material-system.slang

This returnsl nullptr:

        IROffsetDecoration *const fieldOffsetDecoration =
            field->findDecoration<IROffsetDecoration>();

`slang-test -bindir %BINDIR% -category full -api wgpu tests/autodiff/existential-1.slang`

### PR DESCRIPTION
**Pull request by aleino-nv**
Struct legalization removing fields not representable in memory should transfer offset decorations in case the struct has already had offsets calculated.

Closes #5264.

### CODE REVIEW

File: source/slang/slang-legalize-types.cpp [hunk 1]
```diff
@@ -487,8 +487,13 @@ struct TupleTypeBuilder
...
                 // TODO: shallow clone of modifiers, etc.
-
-                builder->createStructField(ordinaryStructType, ee.fieldKey, fieldType);
+                IRStructField* originalField = findStructField(originalStructType, ee.fieldKey);
+                IRStructField* newField =
+                    builder->createStructField(ordinaryStructType, ee.fieldKey, fieldType);
+                // In case the original struct had offsets calculated, transfer those as well.
+                // The original offsets should still be valid, since we only skip fields of types
+                // that aren't representable in memory.
+                originalField->transferChildrenOfTypeTo<IROffsetDecoration>(newField);
```
Comment (csyonghe) [inline source/slang/slang-legalize-types.cpp:496]: Just transfer all decorations over.

File: source/slang/slang-ir.h [hunk 2]
```diff
@@ -780,8 +793,20 @@ struct IRInst
     void removeOperand(Index index);
 
     /// Transfer any decorations of this instruction to the `target` instruction.
+    // TODO: rewrite as transferDecorationsTo(target) ->
```
Comment (csyonghe) [inline source/slang/slang-ir.h:796]: Superfluous comment.

File: source/slang/slang-ir.h [hunk 3]
```diff
@@ -593,6 +593,19 @@ struct IRInst
     IRDecoration* getLastDecoration();
     IRInstList<IRDecoration> getDecorations();
 
+    // Step to the next child of a certain type, T
```
Comment (csyonghe) [inline source/slang/slang-ir.h:596]: Don’t need this, just copy all decorations.

File: source/slang/slang-ir.h (hunk 2 above)
Comment (aleino-nv) [reply source/slang/slang-ir.h:796]: Done.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**